src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,237 @@
+    friend class producer_state;
+};
+
+// A cached buffer of requests, the requests can be in progress / finished.
+// A request is promoted from inflight to finished once it is applied in the
+// log.
+//
+// We retain a maximum of `requests_cached_max` finished requests.
+// Kafka clients only issue requests in batches of 5, the queue is fairly small
+// at all times.
+class requests {
+public:
+    result<request_ptr> try_emplace(
+      seq_t first, seq_t last, model::term_id current, bool reset_sequences);
+
+    bool stm_apply(const model::batch_identity& bid, kafka::offset offset);
+
+    void shutdown();
+
+    friend std::ostream& operator<<(std::ostream&, const requests&);
+
+private:
+    static constexpr int32_t requests_cached_max = 5;
+    // chunk size of the request containers to avoid wastage.
+    static constexpr size_t chunk_size = std::bit_ceil(
+      static_cast<unsigned long>(requests_cached_max));
+    bool is_valid_sequence(seq_t incoming) const;
+    std::optional<request_ptr> last_request() const;
+    ss::chunked_fifo<request_ptr, chunk_size> _inflight_requests;
+    ss::chunked_fifo<request_ptr, chunk_size> _finished_requests;
+    friend producer_state;
+};
+
+/// Encapsulates all the state of a producer producing batches to
+/// a single raft group. At init, the producer registers itself with
+/// producer_state_manager that manages the lifecycle of all the
+/// producers on a given shard.
+class producer_state {
+public:
+    producer_state(
+      producer_state_manager& mgr,
+      model::producer_identity id,
+      raft::group_id group,
+      ss::noncopyable_function<void()> hook)
+      : _id(id)
+      , _group(group)
+      , _parent(std::ref(mgr))
+      , _last_updated_ts(ss::lowres_system_clock::now())
+      , _post_eviction_hook(std::move(hook)) {
+        register_self();
+    }
+
+    producer_state(const producer_state&) = delete;
+    producer_state& operator=(producer_state&) = delete;
+    producer_state(producer_state&&) noexcept = delete;
+    producer_state& operator=(producer_state&& other) noexcept = delete;
+    ~producer_state() noexcept { deregister_self(); }
+
+    friend std::ostream& operator<<(std::ostream& o, const producer_state&);
+
+    /// Runs the passed async function under the op_lock scope.
+    /// Additionally does the following
+    /// - de-registers self from the manager as a pre-hook
+    /// - re-registers with the manager back after the function
+    ///   completes
+    /// This helps the manager implement a lock-free eviction approach.
+    /// - A producer_state with an inflight request is not evicted, because it
+    ///   is no longer in the list of producers tracked.
+    /// - Re-registration helps the manager track LRU-ness, re-registration
+    ///   effectively puts this producer at the end of the queue when
+    ///   considering candidates for eviction.
+    template<AcceptsUnits AsyncFunc>
+    auto run_func(AsyncFunc&& func) {
```

Comment:
`run_func` is very generic name and it hides the fact that it uses a lock; this information is crucial for a reader because every lock is a liability: they cause problems with performance and availability. `with_lock` or `run_with_lock` is a better name